Configuration Cleanup for beta release (#495)
open-telemetry/opentelemetry-dotnet-instrumentation: docs/config.md

> | Environment variable | Description | Default |

Reviewer: ```suggestion
| Environment variable | Description | Default value |
```

Reviewer: For consistency, please change all the "Default" and "Value" columns to "Default value" (if that is indeed their meaning). For example, see the tables from line 127 onwards.

Author: Done in 6d5921ac1ad1b0e9602f1055f5fa527f1e8711ec 👍🏼

|-|-|-|
| `OTEL_DOTNET_AUTO_HOME` | Installation location. | `true` |
| `OTEL_DOTNET_AUTO_HOME` | Installation location. |  |
| `OTEL_DOTNET_AUTO_ENABLED` | Enables the tracer. | `true` |
| `OTEL_DOTNET_AUTO_INCLUDE_PROCESSES` | Names of the executable files that the profiler can instrument. Supports multiple comma-separated values, for example: `MyApp.exe,dotnet.exe`. If unset, the profiler attaches to all processes by default. |  |
| `OTEL_DOTNET_AUTO_EXCLUDE_PROCESSES` | Names of the executable files that the profiler cannot instrument. Supports multiple comma-separated values, for example: `ReservedProcess.exe,powershell.exe`. The list is processed after `OTEL_DOTNET_AUTO_INCLUDE_PROCESSES`. If unset, the profiler attaches to all processes by default. |  |

> | `OTEL_DOTNET_AUTO_AZURE_APP_SERVICES` | Set to indicate that the profiler is running in the context of Azure App Services. | `false` |

Reviewer 2: This is an interesting one, because it is used to toggle on some process exclusion logic that applies to azure environments. Maybe some of these things should not be called out here, but belong in some other document? Maybe it makes sense to have a document related to azure at some point?

Reviewer 3: Perhaps we should track the "non-documented" or "internal" env vars in different document so we can keep the record of the intended usage and later we move them to the proper markdown giving it the proper visibility.

Reviewer: I suggest moving those to a `dev` or `internal` folder within `docs`.

Author: Good idea, added to `/docs/internal/config.md`

| `OTEL_DOTNET_AUTO_DISABLED_INSTRUMENTATIONS` | Comma-separated list of source and bytecode instrumentations you want to disable. | |
| `OTEL_DOTNET_AUTO_{0}_ENABLED` | Configuration pattern for enabling or disabling specific bytecode. For example, to disable GraphQL instrumentation, set the `OTEL_TRACE_GraphQL_ENABLED` environment variable to `false`. | `true` |
| `OTEL_DOTNET_AUTO_DOMAIN_NEUTRAL_INSTRUMENTATION` | Whether to intercept method calls when the caller method is inside a domain-neutral assembly. Useful when instrumenting IIS applications. | `false` |
| `OTEL_DOTNET_AUTO_CLR_DISABLE_OPTIMIZATIONS` |  Set to `true` to disable all JIT optimizations. | `false` |
| `OTEL_DOTNET_AUTO_CLR_ENABLE_INLINING` | Set to `false` to disable JIT inlining. | `true` |

> | `OTEL_DOTNET_AUTO_CLR_ENABLE_NGEN` | Set to `false` to disable NGEN images. | `true` |

Reviewer 2: Would it make sense to have these defined in either an advanced scenarios document(s) or some troubleshooting document? It's probably more important the explain why these options exist, than to document them here.

Reviewer: Agree!

Author: Moved to `docs/internal/config.md`

| Environment variable | Description | Default |
|-|-|-|
| `OTEL_TRACES_EXPORTER` | Traces exporter to be used. The value can be one of the following: `zipkin`, `jaeger`, `otlp`, `none`. | `otlp` |

### Jaeger

> To enable the Jaeger exporter, set `OTEL_TRACES_EXPORTER=jaeger`. To customize the Jaeger exporter via environment variables, see the [Jaeger exporter documentation](https://github.com/open-telemetry/opentelemetry-dotnet/tree/core-1.2.0/src/OpenTelemetry.Exporter.Jaeger#environment-variables).

Reviewer 2: It's easier for us to maintain the documentation by just referring to the SDK docs, however there were concerns previously raised about the user experience and not being able to get the desired information as quickly. Maybe we should do a little of both, highlight the most important settings, and also refer to the SDK documentation?

Reviewer 2: I think that it makes it easier to understand by breaking it up like you've done.

> To enable the Jaeger exporter, set `OTEL_TRACES_EXPORTER=jaeger`. To customize the Jaeger exporter via environment variables, see the [Jaeger exporter documentation](https://github.com/open-telemetry/opentelemetry-dotnet/tree/core-1.2.0/src/OpenTelemetry.Exporter.Jaeger#environment-variables).

Reviewer: ```suggestion
To enable the Jaeger exporter, set the `OTEL_TRACES_EXPORTER` environment variable to `jaeger`. To customize the Jaeger exporter using environment variables, see the [Jaeger exporter documentation](https://github.com/open-telemetry/opentelemetry-dotnet/tree/core-1.2.0/src/OpenTelemetry.Exporter.Jaeger#environment-variables).
```

### OTLP

> To enable the OTLP exporter, set `OTEL_TRACES_EXPORTER=otlp`. To customize the OTLP exporter via environment variables, see the [OTLP exporter documentation](https://github.com/open-telemetry/opentelemetry-dotnet/tree/core-1.2.0/src/OpenTelemetry.Exporter.OpenTelemetryProtocol#environment-variables).

Reviewer 3: We need to call attention around here that auto instrumentation defaults to `http/protobuf` instead of `grpc` - this is different than the default behavior of the package. We opted for that in order to avoid the default of OTEL_DOTNET_AUTO_HTTP2UNENCRYPTEDSUPPORT_ENABLED to be true.

That we perhaps can refine the section of OTLP in a separate PR, but, before beta. I want to review the usage and defaults after we upgrade to v1.2.0 of the SDK and  exporters.

> To enable the OTLP exporter, set `OTEL_TRACES_EXPORTER=otlp`. To customize the OTLP exporter via environment variables, see the [OTLP exporter documentation](https://github.com/open-telemetry/opentelemetry-dotnet/tree/core-1.2.0/src/OpenTelemetry.Exporter.OpenTelemetryProtocol#environment-variables).

Reviewer: ```suggestion
To enable the OTLP exporter, set the `OTEL_TRACES_EXPORTER` environment variable to `otlp`. To customize the OTLP exporter using environment variables, see the [OTLP exporter documentation](https://github.com/open-telemetry/opentelemetry-dotnet/tree/core-1.2.0/src/OpenTelemetry.Exporter.OpenTelemetryProtocol#environment-variables).
```

  to use the `grpc` OTLP exporter protocol. For example, by adding 
  `<PackageReference Include="Grpc.Net.Client" Version="2.32.0" />` to the `.csproj` file.
- On .NET Framework, the `grpc` OTLP exporter protocol is not supported.

### Zipkin

> To enable the Zipkin exporter, set `OTEL_TRACES_EXPORTER=zipkin`. To customize the Zipkin exporter via environment variables, see the [Zipkin exporter documentation](https://github.com/open-telemetry/opentelemetry-dotnet/tree/core-1.2.0/src/OpenTelemetry.Exporter.Zipkin#configuration-using-environment-variables).

Reviewer: ```suggestion
To enable the Zipkin exporter, set the `OTEL_TRACES_EXPORTER` environment variable to `zipkin`. To customize the Zipkin exporter using environment variables, see the [Zipkin exporter documentation](https://github.com/open-telemetry/opentelemetry-dotnet/tree/core-1.2.0/src/OpenTelemetry.Exporter.Zipkin#configuration-using-environment-variables).
```

To perform bytecode instrumentation, configure the OpenTelemetry .NET
Automatic Instrumentation as a .NET CLR Profiler. The CLR uses the following
environment variables to set up the profiler. 
environment variables to set up the profiler. See [.NET Runtime Profiler Loading](https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/profiling/Profiler%20Loading.md) for more information.

### .NET Framework
| Environment variable | Description | Value |
|-|-|-|
| `COR_ENABLE_PROFILING` | Enables the profiler. | `1` |
| `COR_PROFILER` | CLSID of the profiler. | `30000` (ms) |
| `COR_PROFILER_PATH` | Path to the profiler. | `%InstallationLocation%/OpenTelemetry.AutoInstrumentation.Native.dll` |
| `COR_PROFILER_PATH_32` | Path to the 32-bit profiler. Bitness-specific paths take precedence over generic paths. | `%InstallationLocation%/win-x86/OpenTelemetry.AutoInstrumentation.Native.dll` |
| `COR_PROFILER_PATH_64` | Path to the 64-bit profiler. Bitness-specific paths take precedence over generic paths. | `%InstallationLocation%/win-x64/OpenTelemetry.AutoInstrumentation.Native.dll` |

> > Notice that .NET Framework uses the `COR_` prefix instead of `CORECLR_`.

Reviewer 3: Can we keep a single table and perhaps call more attention to the different prefixes?

Author: Tried to address this in a5c96d92f0460ff7a67b90cc3846a6671024ce80, let me know what you think

> ### .NET Core (including .NET 5+)

Reviewer: ```suggestion
### .NET Core, including .NET 5 and higher
```

Author: Thanks for the suggestion. In an effort to consolidate tables, this may no longer be relevant. But if it becomes relevant again I can apply this suggestion

To perform bytecode instrumentation, configure the OpenTelemetry .NET
Automatic Instrumentation as a .NET CLR Profiler. The CLR uses the following
environment variables to set up the profiler. See [.NET Runtime Profiler Loading](https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/profiling/Profiler%20Loading.md) for more information.

> | .NET Framework environment variable | .NET Core environment variable | Description | Required value |

Reviewer 3: 👍
